Tickets/DM-12605: Add draft from Data Facility

hsinfang opened the pull request.

Commit:
Add a first draft on DRP-00-05 test case for internal review

Commit:
Clarify software branches used to avoid commas in the patch IDs

Commit:
Add more requirements for future milestones

Commit:
Reorganize the Procedure in DRP-00-05 test case

Commit:
Remove requirements that are more for DAC, DBB, SuperTask, etc

Commit:
Replace test items with original requirement texts and reorder

MichelleGower committed:
Modified document.

MichelleGower committed:
Update the lsst-texmf submodule to more recent version

jdswinbank commented on cases/drp-00-05.tex:
```diff
@@ -3,16 +3,24 @@ \subsection{DRP-00-05: Execution of the DRP Science Payload by the DRP Execution
+This test will check that the DRP Science Payload can be executed using the current state-of-the-art execution service provided by the LSST Data Facility.
 
-\begin{itemize}
+Specifically, this will check if the following requirements are fulfilled:
```
We should tweak the wording here to clarify that we're only demonstrating that these requirements are fulfilled _in the context of this particular production_.

We are not claiming that by executing this DESDM-specific test case we will have demonstrated that this requirement is fulfilled by the system which will be delivered to the operations team.

jdswinbank commented on cases/drp-00-05.tex:
```diff
@@ -3,16 +3,24 @@ \subsection{DRP-00-05: Execution of the DRP Science Payload by the DRP Execution
-  state-of-the-art execution service provided by the LSST Data Facility.}
+\begin{itemize}
+  \item{For each Coadded Image, DMS shall store: the list of input images and the pipeline parameters, including software versions, used to derive it, and a sufficient set of metadata attributes for users to re-create them in whole or in part (DMS-REQ-0106).}
+  \item{To reduce the likelihood for confusion, all IDs shall be unique across databases and database versions, other than those corresponding to uniquely identifiable entities (i.e., IDs of exposures) (DMS-REQ-0292).}
```
This is a good example of a requirement which we cannot fully demonstrate compliance with in this test case. The DMSR specifically calls out multiple separate data releases sharing no identical Object etc IDs; since we only have one (mini) data release in this test, we can't show that we're meeting that requirement.

(Of course, we can show that there are no duplicate Object IDs in the particular data release produced, but that's just a subset of the full requirement.)

hsinfang replied:
We thought of that too.  Many DMSR items are boarder than this test case is about. We listed those we find most relevant although part of the requirements ought to be fulfilled by some other parts of the system. 

I copied the original DMSR directly here. I'm tempted to delete them as they are duplicate as it is. Opinions?

jdswinbank replied:
I agree that simply copying the DMSR here is of dubious usefulness.

My interpretation is that the "test items" should list the products/components under test, rather than the requirements which refer to them. So, for example, the test items might be "processing middleware" or "installation procedures" or a particular data product, rather than "DMS-REQ...".

This makes more sense in the context of a verification matrix: a requirement is satisfied by passing a number of test cases, each of which tests just some subset of the components & products which are necessary to meet that requirement.

jdswinbank commented on cases/drp-00-05.tex:
```diff
@@ -3,16 +3,24 @@ \subsection{DRP-00-05: Execution of the DRP Science Payload by the DRP Execution
+  \item{To reduce the likelihood for confusion, all IDs shall be unique across databases and database versions, other than those corresponding to uniquely identifiable entities (i.e., IDs of exposures) (DMS-REQ-0292).}
+  \item{A Dataset may consist of one or more pixel images, a set of records in a file or database, or any other grouping of data that are processed or produced as a logical unit. The DMS shall be able to identify and retrieve complete, consistent datasets for processing (DMS-REQ-0293).}
+  \item{The DMS shall process all requested datasets until either a successful result is recorded or a permanent failure is recognized. If any dataset is processed, in part or in whole, more than once, only one of the wholly processed results will be recorded for further processing.
+(DMS-REQ-0294).}
```
Unless you plan to artificially induce duplicate processig (and I don't think the test procedure below describes doing so), you aren't testing the second part of this requirement (except by chance, I suppose).

MichelleGower replied:
Actual testing of this requirement was not planned in this milestone.   In order to verify DMS-REQ-0294 in some future test, yes we would have to resubmit the same pipeline multiple times (with and without errors).    I assume since it is a requirement, someone will have to someday do this test.    It sounds like we either need to temporarily remove this requirement from this document or modify the procedure to include the steps.

jdswinbank commented on cases/drp-00-05.tex:
```diff
@@ -3,16 +3,24 @@ \subsection{DRP-00-05: Execution of the DRP Science Payload by the DRP Execution
+(DMS-REQ-0294).}
+  \item{The DMS shall provide software to orchestrate execution of productions, including deploying pipelines on a computing platform (DMS-REQ-0302).}
+  \item{The DMS shall provide software to monitor execution of pipelines in real time (DMS-REQ-0303).}
+  \item{The DMS shall provide software to detect faults in pipeline execution and recover when possible.
```
Are you testing this? I think you'd need to induce a fault and demonstrate recovery, and I don't think the test procedure includes that.

jdswinbank replied:
(Of course, if a fault happens to occur spontaneously, that would demonstrate this, but I assume we can't rely on that!)

MichelleGower replied:
(We accidentally have the spontaneously occurring faults during our runs for this milestone due to the networking problems on LSST-VC.)  Same answer as above, we either need to temporarily drop the requirement or modify the test procedure.

jdswinbank commented on cases/drp-00-05.tex:
```diff
@@ -26,32 +34,149 @@ \subsubsection{Intercase dependencies}
 
-\paragraph{Software}
+\paragraph{Software}\label{sec:software}
+The LSST DM Software Stack and the DESDM Framework packages will be pre-installed.
```
Are you using the results of DRP-00-00 to provide this, or is it a separate installation?

jdswinbank replied:
For DESDM, I don't think we have a test case describing installation and I don't think it's strictly necessary to produce one for now. However, assuming such a test case is not provided, we should at least provide a reference to a document describing the DESDM installation procedure.

hsinfang replied:
For LDM-503-2 it's a separate installation.  In the future it may be a same installation.

MichelleGower replied:
Because the software is spread across 3 different sources, the installation process is currently an interesting manual process.    As @hsinfang stated, if/when the DESDM software base has been chosen to become part of the LSST software stack, it would fall under DRP-00-00.  This decision is slated for late 2018/very early 2019 AFAIK.   We had planned on writing up this manual process post-Nov after having more experience installing on machines from scratch.    How important is it that this temporary installation process is written down for this version of the document?

jdswinbank commented on cases/drp-00-05.tex:
```diff
@@ -26,32 +34,149 @@ \subsubsection{Intercase dependencies}
+
+For LDM-503-2, the three tracts of the Hyper Suprime-Cam "RC" dataset, as defined in DMTR-31, will be used.
+The calibration dataset will be the 20170105 version, as in DMTR-31.   Raw files known to fail processCcd 
+should be blacklisted.
```
s/should/will/ :-)

jdswinbank commented on cases/drp-00-05.tex:
```diff
@@ -26,32 +34,149 @@ \subsubsection{Intercase dependencies}
-\end{note}
+\subsubsection{Input specification}\label{sec:input}
+
+One to three tracts of the Hyper Suprime-Cam dataset will be used along with appropriate calibration datasets.
```
Why "one to three"? I don't think that it's strictly necessary to specify for the purposes of this document, but I do wonder why you don't.

hsinfang replied:
Changed to "A small number of selected tracts"

jdswinbank commented on cases/drp-00-05.tex:
```diff
@@ -26,32 +34,149 @@ \subsubsection{Intercase dependencies}
-\end{note}
+The output data products will be available from the Data Backbone.
+
+For LDM-503-2, the output data products will be available on the LSST-VC via a shared filesystem and data discovery is done via SQL queries against the Oracle database.
```
Is it implicit in the above that the data products will not be made available in a Butler repository, but simply dumped to the filesystem?

(I don't see that as a problem for the test; I'm just curious)

hsinfang replied:
Making the data products accessible via Butler is not a goal for Nov, and may need the Butler team to write a tool to do so properly.

MichelleGower replied:
It was not clear that anyone would actually be interested in using these files as the test definition produces not current science-ready products.    Upcoming work will involve producing outputs where scientists will be asked to inspect them.    It is not clear whether a prototype new Butler would exist by then that could access the prototype Data Backbone without a separate tool as would be needed by the current Butler.

hsinfang replied:
I managed to make a limited set of output products assessable via Butler, using a hand-made repo config. So if anyone would be interested in using the outputs I could make an entry repo in /project/ or somewhere.

jdswinbank commented on cases/drp-00-05.tex:
```diff
@@ -26,32 +34,149 @@ \subsubsection{Intercase dependencies}
+    \begin{enumerate}
+        \item{Ask the Data Backbone about the files generated by each tract's execution through the pipeline.}
+        \item{Verify the physical and location information in Data Backbone DB matches the Data Backbone filesystem and vice-versa.} 
+    \end{enumerate}
```
Do we do any check on the correctness of the Data Backbone, as opposed to just is consistency?

That is, my reading of the above is that it asks the Data Backbone what files were generated, and then it checks those files were consistently recorded on the filesystem and in the database. However, we should probably also check that list of files the Data Backbone things were generated is correct.

(If the DBB reports that no files were generated, and you verify that no files are recorded on the filesystem or in the database, you would seem to pass the above test without actually doing anything very useful!)

MichelleGower replied:
The first part would verify that the "correct" files were generated by asking the DB, but it assumes the DB matches the filesystem.    The second part verifies that the filesystem matches the DB.

jdswinbank commented on cases/drp-00-05.tex:
```diff
@@ -26,32 +34,149 @@ \subsubsection{Intercase dependencies}
+    \begin{enumerate}
+        \item{Generically check that each file can be linked with the task and processing attempt that created it via the Data Backbone.}
+        \item{Generically check that the inputs for each step were saved to the Oracle database.}
+        \item{Specifically check conditions of DMS-REQ-0106.}
```
What is "specifically" checked here? 0106 is itself quite vague ("a sufficient set of metadata attributes for users to re-create them in whole or in part") — are you going to enumerate the set of metadata which will be checked, or are you actually going to demonstrate the recreation of data products from metadata?

hsinfang replied:
I deleted this item as I think the two previous items are already sufficient. @MichelleGower please push back if you disagree

jdswinbank commented on cases/drp-00-05.tex:
```diff
@@ -26,32 +34,149 @@ \subsubsection{Intercase dependencies}
+
+\paragraph{First-order Science QA}
+\begin{enumerate}
+  \item{Production Scientist will perform first-order science quality assurance.}
```
What is this "first-order science quality assurance"?

Is it necessary to include it in this test case? I think it is orthogonal to the stated goals of the test.

jdswinbank commented on cases/drp-00-05.tex:
```diff
@@ -26,32 +34,149 @@ \subsubsection{Intercase dependencies}
+\paragraph{First-order Science QA}
+\begin{enumerate}
+  \item{Production Scientist will perform first-order science quality assurance.}
+  \item{These tests were not scheduled as part of LDM-503-2, but will be specified and used in future work.}
```
...ah. Maybe this answers my question above. I think if it's not necessary to do this work, we shouldn't record it here. (That doesn't mean it shouldn't be done if it's useful for other purposes, just that we don't need to confuse things by mentioning it.)

MichelleGower replied:
The goal in putting it here was a placeholder since the LDF plans to provide first-order science QA testing as part of future test runs of pipeline processing.

jdswinbank commented on cases/drp-00-05.tex:
```diff
@@ -26,32 +34,149 @@ \subsubsection{Intercase dependencies}
+\begin{enumerate}
+  \item{Production Scientist will perform first-order science quality assurance.}
+  \item{These tests were not scheduled as part of LDM-503-2, but will be specified and used in future work.}
+  \item{Any failure of non-science QA can cause an attempt to be removed from the campaign tag, marked as a bad attempt, and be resubmitted.} 
```
I don't understand why this is under the "First-order Science QA" heading.

MichelleGower replied:
"non-science QA" was a copy and paste error (which Hsin-Fang is already fixing).    It should read just "science QA".   I was less sure of whether a science QA error would trigger a resubmit (with configuration tweaking) or just always end in immediate test failure.  So the resubmit part could be removed if we're not sure.

Commit:
Update case DRP-00-05 for review comments

Commit:
Correct typo

MichelleGower committed:
Improve text re Batch Processing Service

Remove text about processing attempt retries/resubmits and
anything across multiple attempts.  Added testbed description.
Made various formatting changes.

Commit:
Fix task config specification and section labels
